src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -254,34 +279,59 @@ module Make (Inputs : Inputs_intf.S) = struct
 
     let merkle_path_at_addr_exn t address =
       assert_is_attached t ;
-      let parent_merkle_path =
-        Base.merkle_path_at_addr_exn (get_parent t) address
-      in
-      fixup_merkle_path t parent_merkle_path address
+      match self_merkle_path t address with
+      | Some path ->
+          path
+      | None ->
+          let parent_merkle_path =
+            Base.merkle_path_at_addr_exn (get_parent t) address
+          in
+          fixup_merkle_path t parent_merkle_path address
 
     let merkle_path_at_index_exn t index =
-      assert_is_attached t ;
-      let address = Addr.of_int_exn ~ledger_depth:t.depth index in
-      let parent_merkle_path =
-        Base.merkle_path_at_addr_exn (get_parent t) address
-      in
-      fixup_merkle_path t parent_merkle_path address
+      merkle_path_at_addr_exn t (Addr.of_int_exn ~ledger_depth:t.depth index)
 
     let merkle_path t location =
-      assert_is_attached t ;
-      let address = Location.to_path_exn location in
-      let parent_merkle_path = Base.merkle_path (get_parent t) location in
-      fixup_merkle_path t parent_merkle_path address
+      merkle_path_at_addr_exn t (Location.to_path_exn location)
 
     let merkle_path_batch t locations =
       assert_is_attached t ;
-      let addresses = List.map ~f:Location.to_path_exn locations in
-      let parent_merkle_paths =
-        Base.merkle_path_batch (get_parent t) locations
+      let self_merkle_paths_rev =
+        List.rev_map locations ~f:(fun location ->
+            let address = Location.to_path_exn location in
+            match self_merkle_path t address with
+            | Some path ->
+                Either.First path
+            | None ->
+                Either.Second (location, address) )
+      in
+      let parent_merkle_paths_rev =
+        let parent_locations_rev =
+          List.filter_map self_merkle_paths_rev ~f:(function
+            | Either.First _ ->
+                None
+            | Either.Second (location, _) ->
+                Some location )
+        in
+        if List.is_empty parent_locations_rev then []
+        else Base.merkle_path_batch (get_parent t) parent_locations_rev
+      in
+      let rec recombine self_merkle_paths_rev parent_merkle_paths_rev acc =
+        match (self_merkle_paths_rev, parent_merkle_paths_rev) with
+        | [], [] ->
+            acc
+        | Either.First path :: self_merkle_paths_rev, parent_merkle_paths_rev ->
+            recombine self_merkle_paths_rev parent_merkle_paths_rev (path :: acc)
+        | ( Either.Second (_, address) :: self_merkle_paths_rev
+          , path :: parent_merkle_paths_rev ) ->
+            let path = fixup_merkle_path t path address in
+            recombine self_merkle_paths_rev parent_merkle_paths_rev (path :: acc)
+        | _ :: _, [] ->
```

Comment:
Refining this case as:
```suggestion
        | Second _ :: _, [] ->
```
will make it clearer why we think this cannot happen. Although a comment explaining invariants we rely on here would still be very welcome.